pkg/defaultcomponents/defaults_test.go

```diff
@@ -101,4 +103,47 @@ func TestComponents(t *testing.T) {
 	assert.NotNil(t, processors["groupbytrace"])
 	assert.NotNil(t, processors["tail_sampling"])
 	assert.NotNil(t, processors["k8sattributes"])
+
+	// Ensure that the components behind feature gates aren't included
+	assert.Nil(t, receivers["filelog"])
+	assert.Nil(t, exporters["awscloudwatchlogs"])
+	assert.Nil(t, extensions["file_storage"])
+}
+
+func TestEnabledFeatureGate(t *testing.T) {
+	testCases := []struct {
+		featureName string
+		expectedLen int
+	}{
+		{"adot.receiver.filelog", receiversCount + 1},
+		{"adot.exporter.awscloudwatchlogs", exportersCount + 1},
+		{"adot.extension.file_storage", extensionsCount + 1},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.featureName, func(t *testing.T) {
+			err := featuregate.GlobalRegistry().Set(tc.featureName, true)
+			assert.NoError(t, err)
+
+			factories, err := Components()
+			assert.NoError(t, err)
+
+			switch tc.featureName {
+			case "adot.receiver.filelog":
+				receivers := factories.Receivers
+				assert.Len(t, receivers, tc.expectedLen)
+				assert.NotNil(t, receivers["filelog"])
+
+			case "adot.exporter.awscloudwatchlogs":
+				exporters := factories.Exporters
+				assert.Len(t, exporters, tc.expectedLen)
+				assert.NotNil(t, exporters["awscloudwatchlogs"])
+
+			case "adot.extension.file_storage":
+				extensions := factories.Extensions
+				assert.Len(t, extensions, tc.expectedLen)
+				assert.NotNil(t, extensions["file_storage"])
+			}
```

Comment:
nit: You don't have to resolve it in this PR but I don't believe this is the correct way to write table based tests in go. 

When writing table based tests in go all of your test setup should be done when configuring each test case. When the test case are iterated through the setup and execution should look identical for each test case. You shouldn't need to rely on a switch statement like this. 

It's possible my original comment is wrong and these don't fit cleanly into a table based test. It's also possible that you need to figure out how to make your test cases more generic. 

I'd be happy to go over different ways we could look at creating these test cases in the future. For now they still perform the assertions that we want.

Reply:
Thanks bryan, makes sense